compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1801,21 +1811,65 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// Report an error for a struct field expression when there are invisible fields.
     ///
     /// ```text
-    /// error: cannot construct `Foo` with struct literal syntax due to inaccessible fields
+    /// error: cannot construct `Foo` with struct literal syntax due to private fields
     ///  --> src/main.rs:8:5
     ///   |
     /// 8 |     foo::Foo {};
     ///   |     ^^^^^^^^
     ///
     /// error: aborting due to previous error
     /// ```
-    fn report_inaccessible_fields(&self, adt_ty: Ty<'tcx>, span: Span) {
-        self.tcx.sess.span_err(
+    fn report_private_fields(
+        &self,
+        adt_ty: Ty<'tcx>,
+        span: Span,
+        private_fields: Vec<&ty::FieldDef>,
+        used_fields: &'tcx [hir::ExprField<'tcx>],
+    ) {
+        let field_names = |fields: Vec<Symbol>, len: usize| match &fields
+            .iter()
+            .map(|field| field.to_string())
+            .collect::<Vec<_>>()[..]
+        {
+            _ if len > 6 => String::new(),
+            [name] => format!("`{name}` "),
+            [names @ .., last] => {
+                let names = names.iter().map(|name| format!("`{name}`")).collect::<Vec<_>>();
+                format!("{} and `{last}` ", names.join(", "))
+            }
+            [] => unreachable!(),
+        };
+
+        let mut err = self.tcx.sess.struct_span_err(
             span,
             &format!(
-                "cannot construct `{adt_ty}` with struct literal syntax due to inaccessible fields",
+                "cannot construct `{adt_ty}` with struct literal syntax due to private fields",
             ),
         );
+        let (used_private_fields, remaining_private_fields): (
+            Vec<(Symbol, Span)>,
+            Vec<(Symbol, Span)>,
+        ) = private_fields.iter().partition_map(|field| {
+            match used_fields.iter().find(|used_field| field.name == used_field.ident.name) {
+                Some(used_field) => Either::Left((field.name, used_field.span)),
+                None => Either::Right((field.name, self.tcx.def_span(field.did))),
+            }
+        });
+        let remaining_private_fields_len = remaining_private_fields.len();
+        err.span_labels(used_private_fields.iter().map(|(_, span)| *span), "private field");
+        err.span_note(
+            MultiSpan::from_spans(remaining_private_fields.iter().map(|(_, span)| *span).collect()),
+            format!(
+                "missing field{s} {names}{are} private",
+                s = pluralize!(remaining_private_fields_len),
+                are = pluralize!("is", remaining_private_fields_len),
+                names = field_names(
```

Comment:
Why is this a closure call? Can you inline the body of `field_names`?

Reply:
I originally used this closure elsewhere, but forgot to remove it when I finally didn't need it.